lib/auto_attach.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
+
+"""
+Perform auto-attach operation
+
+On Ubuntu Pro machines, we try to perform an auto-attach operation
+on first boot. This happens through a systemd unit that triggers
+that executes this script on every boot. However, if we detect
+that cloud-init has user data related to ua, we cannot run
+auto-attach here, since cloud-init will drive this operation on
+their side.
+"""
+import logging
+
+import yaml
+
+from uaclient.cli import action_auto_attach, setup_logging
+from uaclient.config import UAConfig
+from uaclient.exceptions import (
+    AlreadyAttachedOnPROError,
+    ProcessExecutionError,
+)
+from uaclient.util import subp, which
+
+
+def check_cloudinit_userdata_for_ua_info():
+    if not which("cloud-init"):
+        return False
+
+    try:
+        userdata, _ = subp(["cloud-init", "query", "userdata"])
```

Comment:
I'm not that familiar with the cloud-init query feature. But I thought we'd be able to do something like:
```suggestion
        userdata, _ = subp(["cloud-init", "query", "userdata.ubuntu_advantage"])
```
Where that only returns data if there is anything set in the ubuntu_advantage module in userdata. This would let us skip parsing the yaml ourselves.

Is something like that possible @blackboxsw ?

Reply:
@orndorffgrant currently `cloud-init query userdata` only returns the raw userdata which could be `#cloud-config` YAML,  but it could also be a `#include-url https://someurl-which-has-cloud-config-stuff` or it could even be a binary blob that cloud-init doesn't react to. So, the expectation of userdata being a structured object that we can dive into can be a problem in some cases.

What I had originally thought was that we could use `cloud-init query merged_cfg.ubuntu_advantage` because the merged config is guaranteed to be rendered based on merged sources. But, I neglected to recall that `merged_cfg` is only merged config from 'system config' which is the aggregated files in /etc/cloud/cloud.cfg and /etc/cloud/cloud.cfg.d/*.
So, merged_config only gives you part of the puzzle. You'll need to query both userdata and merged_cfg from the filesystem somehow.

Long-term I think we want cloud-init to make `merged_cfg` reflect not only /etc/cloud/cloud.cfg* but also the merged userdata and vendordata which affects the configuration of the VM. But, this feature may require a bit of work outside of this cycle because userdata and vendordata in some cases are processed and rendered in the "init network" boot stage instead of init-local when the datasource is initially detected.

At the moment, the best solution to query both userdata and merged_cfg from disk is one of 3 options:
 1. Grab it the merged config direct from the fully realized init config which will merge userdata/vendordata into filesystem config for you
 ```python
from cloudinit.stages import Init
init = Init()
if init.cfg and "ubuntu_advantage" in init.cfg:
    skip_autoattach()
```
 2. Shell out to all instance data and YAML load to check for UA directives:
 ```python
instancedata_yaml, _ = subp(["cloud-init", "query", "--all"])
instance_data = yaml.safe_load(instancedata_yaml)
merged_cfg = instance_data["merged_cfg"]
userdata = yaml.safe_load(instance_data["userdata"])
if "ubuntu_advantage" not in merged_cfg and "ubuntu_advantage" not in userdata:
   skip_autoattach()
```
3. Shell out to all instance data and 'grep' for ubuntu_advantage directives:
````python
instancedata_yaml, _ = subp(["cloud-init", "query", "--all"])
if re.search("ubuntu_advantage:") in instancedata_yaml:
    skip_autoattach()
```

Reply:
Thanks for all the input here @blackboxsw. I like the first idea more to avoid unnecessary shell out. However, `cloud-init` is not one our dependencies at the moment. If we manage to directly import it in the code, should we add it as a dependency ? And I am saying that only for the pro package

Reply:
There is a general desire to move away from having the -pro package at all so I'd rather not add more to it, if we can avoid it.

We can catch an ImportError if the module doesn't exist.

Is there any chance the interface of `cloudinit.stages.Init().cfg` will change in the future though? If it does, it would then break our integration if we rely on it directly.

Reply:
@orndorffgrant `cloudinit.stages.Init().cfg` shouldn't change significantly, and hasn' in the last 6 years. 

cloud-init will keep Init().cfg stable until we can provide UA and others with a succinct representation of all merged_cfg data direct from `/run/cloud-init/instance-data-sensitive.json`. We want the flat files to be our primary API in case we need to change init logic in the future.


Ideally we'd like to steer consumers to read static `/run/cloud-init/instance-data-sensitive.json` `merged_cfg` key instead of re-running the logic that Init().cfg does to remerge all instance data into `.cfg`. We are already [shifting subquity live installer](https://github.com/canonical/subiquity/pull/1352) away from Init().cfg for that reason.

That said, currently /run/cloud-init/instance-data-sensitive.json merged_cfg doesn't have user-data present in merged_cfg and cloud-init will sort publishing user-data/vendor-data in `merged_cfg` before we steer UA away from `Init().cfg`.


@lucasmoura: as discussed separately. There is a potential race in how ua-auto-attach is triggered based on /run/cloud-init/instance-data.json. This file is written before user-data is processed so a race exists where init-network detected datasources (Xenial GCP) will not have rendered the user-data yet and `Init().cfg` may not see it yet. To resolve this to ensure ua-auto-attach.service can't start until user-data is completely visible, we should add an `After=cloud-config.target` to ua-auto-attach.service. I think we can then drop the `ua-auto-attach.path` directives as this will ensure the any applicable instance data to support both `cloud-id` calls and user-data queries via `Init().cfg` is present.

Reply:
Awesome, I have updated the code to address this race condition as discussed. Thanks for all of the investigation you performed for this PR @blackboxsw